src/ILCompiler/src/RdXmlRootProvider.cs

```diff
@@ -196,5 +213,55 @@ private static void RootMethod(IRootingServiceProvider rootProvider, MethodDesc
                 // TODO: Log as a warning
             }
         }
+
+        private static Instantiation GetInstantiationThatMeetsConstraints(Instantiation inst)
+        {
+            TypeDesc[] resultArray = new TypeDesc[inst.Length];
+            for (int i = 0; i < inst.Length; i++)
+            {
+                TypeDesc instArg = GetTypeThatMeetsConstraints((GenericParameterDesc)inst[i]);
+                if (instArg == null)
+                    return default(Instantiation);
+                resultArray[i] = instArg;
+            }
+
+            return new Instantiation(resultArray);
+        }
+
+        private static TypeDesc GetTypeThatMeetsConstraints(GenericParameterDesc genericParam)
+        {
+            GenericConstraints constraints = genericParam.Constraints;
+            if ((constraints & GenericConstraints.NotNullableValueTypeConstraint) != 0)
+                return null;
+
+            TypeDesc result = null;
+            foreach (var c in genericParam.TypeConstraints)
+            {
+                // Can't do multiple type constraints
+                if (result != null)
+                    return null;
+
+                // Constraints in the form of T : U
+                if (c.IsSignatureVariable)
+                    return null;
+
+                // Other potentially complex constraints
+                if (c.HasInstantiation || c.IsArray)
```

Comment:
It means that e.g. `IEnumerable<byte>` constraint will be rejected. It feel too pessimistic.

Should this rather say something like `if (c.ContainsGenericVariables) return null;` ?

Reply:
We don't have means to answer ContainsGenericVariables in the type system besides recursively walking the sig.

This algorithm can get really fancy and I was shooting for something simple that is hard to make a bug in. I added the specific case, but I don't want to take it any further without test coverage.

Reply:
> ContainsGenericVariables in the type system besides recursively walking the sig.

We walk the sig recursively everywhere ... doing it here would not be much more complex than what's there now, and it would be self-explanatory:

```
private static bool ContainsSignatureVariables(TypeDesc type)
{
     if (type.IsSignatureVariable)
         return true;
     
     if (type.IsArray)
         return ContainsSignatureVariables(((ArrayType)type).ElementType);

     foreach (TypeDesc arg in type.Instantiation)
     {
         if (ContainsSignatureVariables(arg))
             return true;
     }

     return false;
}
```

I think this should either have the above; or there should be a comment that it can be `ContainsSignatureVariables` but we choose to handle just some of the cases only.

Reply:
Instead of rooting the thing instantiated over some concrete type, it should be sufficient to just instantiate over `__Canon`, but it doesn't work for some reason right now (I guess because of rhbind limitations that we always need to fight on the ProjectN side that we mirror in CoreRT). The solution to `IEnumerable<byte>` or `IEnumerable<T>` is `__Canon`. We shouldn't need to drill into it or reject because of the `T`. That's why I didn't want to make it any fancier. It doesn't help ASP.NET and it's the wrong direction anyway.

I just don't want to touch the reflection stack to fix that right now because I know it will be trouble when it mirrors over.

Reply:
That's fine ... but I think it would be nice to capture the fact that this is a workaround that should not be needed, and that ideally we should just instantiate over `__Canon`, as comment in the code.

Reply:
I fixed this to use `__Canon`.

The bad thing is I'll still need to dig through the CoreFX test failures before we can merge this.

The good thing is that all of these tests are failing because we didn't run them before (because they were invisible to xUnit). So this pull request is making reflection behave in a more expected manner.